stl/inc/deque

```diff
@@ -871,13 +871,48 @@ public:
         _STL_VERIFY(_Off <= _Mysize(), "deque emplace iterator outside range");
 #endif // _ITERATOR_DEBUG_LEVEL == 2
 
-        if (_Off <= _Mysize() / 2) { // closer to front, push to front then rotate
-            emplace_front(_STD forward<_Valty>(_Val)...);
-            _STD rotate(begin(), _Next_iter(begin()), begin() + static_cast<difference_type>(1 + _Off));
-        } else { // closer to back, push to back then rotate
-            emplace_back(_STD forward<_Valty>(_Val)...);
-            _STD rotate(begin() + static_cast<difference_type>(_Off), _Prev_iter(end()), end());
+        constexpr bool _Uses_move_assignment =
+            disjunction_v<is_nothrow_move_assignable<value_type>, negation<is_copy_assignable<value_type>>>;
+
+        _Orphan_all();
+        if (_Off == 0) { // at the beginning
+            _Emplace_front_internal(_STD forward<_Valty>(_Val)...);
+        } else if (_Off <= _Mysize() / 2) { // closer to front, push to front then rotate
+            auto& _Al = _Getal();
+            _Alloc_temporary2<_Alty> _Obj(_Al, _STD forward<_Valty>(_Val)...); // handle aliasing
+
+            auto _Old_unwrapped_begin = _Unchecked_begin();
+            auto _Old_unwrapped_where = _Old_unwrapped_begin + static_cast<difference_type>(_Off);
+

```

Comment:
I think we'd better acquire the iterators after doing `_Emplace_front_internal`. Otherwise, this relies on that unchecked iterator will implementationaly not be invalidated by `_Emplace_front_internal`.

I find it very hard to reason about its safety. Unchecked iterators has an internal offset, which does not share the same meaning as `_Off` as in `deque[_Off]`. And `_Emplace_front_internal` calls `_Growmap` which can also mess with `_Myoff()` etc(which will be the internal offset for `begin()` (the usage of `off` is just too chaotic)... Even finally safe, I think this is a heavy reliance, as what's going on here is far from obvious.

Reply:
In the future (may be near or in vNext), I think the iterator type used here should be changed to store a beginning `_Map` pointer and an offset in order to reduce indirection. Such an iterator can definitely be invalidated by `_Growmap`.

Reply:
I guess this will not work, as deque relies on ring buffer. An iterator needs 1.mapptr 2.offset to access element, but also needs 3.mapsize to correctly perform iterator arithmetic.